Implement arg_name and USE_ARGS_POSITIONAL

sirosen opened the pull request:
These features allow `use_args` to pass parsed data as keyword arguments without expanding, as `as_kwargs=True`/`use_kwargs` does. This allows for better control and clarity, especially in use-cases where decorators are stacked or may be abstracted by a framework or other tooling.

``arg_name`` functions to set the keyword argument name used for parsed data (rather than appending to positional args). ``USE_ARGS_POSITIONAL = False`` on a parser class enables implicit/automatic values for ``arg_name``, making the default behavior ``{location}_args``. With this set, for example `location="json"` will pass data as `json_args`.

---

These changes and the associated docs address most of #830, but they're intentionally geared towards supporting that use case in v8 and then changing the default behavior to clean up in v9.

sirosen committed:
Implement arg_name and USE_ARGS_POSITIONAL

These features allow `use_args` to pass parsed data as keyword
arguments without expanding, as `as_kwargs=True`/`use_kwargs` does.
This allows for better control and clarify, especially in use-cases
where decorators are stacked or may be abstracted by a framework or
other tooling.

``arg_name`` functions to set the keyword argument name used for
parsed data (rather than appending to positional args).
``USE_ARGS_POSITIONAL = False`` on a parser class enables
implicit/automatic values for ``arg_name``, making the default
behavior ``{location}_args``. With this set, for example
`location="json"` will pass data as `json_args`.

lafrech commented:
Thanks @sirosen for tackling this.

@greyli might want to have a look as well.

greyli commented:
LGTM. Thanks for the quick action!

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

sirosen commented:
I'd like to release this soon along with some more typing refinements (some of which I'm working on right now). It's been a while since our last release and I think this is a good improvement.

I've reviewed my own work with the benefit of time and space, and I'm relatively happy with the implementation.
@lafrech, if you have some time available, would you be willing to give this a final review and :+1: / :-1: it?

lafrech commented on docs/advanced.rst:
```diff
@@ -656,6 +655,83 @@ To reduce boilerplate, you could create shortcuts, like so:
+
+For ``use_args``, the result is that sometimes it is non-obvious which order
+arguments will be passed in. Consider the following two example snippets, one
+of which are nearly identical:
```
> one of which are nearly identical

I'm not a native speaker but I find this sentence a bit strange. What about

> Consider the following nearly identical example snippets:

sirosen replied:
Good catch! Thanks!

I think you're right to find it strange.
"one of which are" isn't grammatical as far as I know. I'm guessing that this was an editing error of mine and that my intent was to make it read
> Consider the following two example snippets, which are nearly identical:

I prefer the phrasing you proposed, "Consider the following nearly identical example snippets", so I'll go with that.
It's a more direct construction and, IMO, easier to read.

lafrech reviewed (commented):
Sorry about the delay, again.

Looks well thought and implemented.

What about the case of the decorator applied twice for the same location with `Schema(partial=True)`? I suppose this would be a legit case. This can work by passing explicit names but if none are passed we should probably raise an error. I didn't give it much thought but I'm not sure we can easily detect this at init.

sirosen commented:
> What about the case of the decorator applied twice for the same location with `Schema(partial=True)`? I suppose this would be a legit case. This can work by passing explicit names but if none are passed we should probably raise an error. I didn't give it much thought but I'm not sure we can easily detect this at init.

That's an interesting use-case. I'll see if I can do something to make this detectable.

I think it would be nice to raise an error when the decorators are evaluated, if possible. It would probably (I want to verify) raise an error at runtime.
If I can't figure it out, we could leave it as is in the current implementation but document the behavior.

sirosen committed:
Make arg_name usage detect conflicts

When `arg_name` is used explicitly or implicitly via the class
setting, the values being passed are stored in a webargs-owned
attribute on the decorated callable (`__webargs_argnames__`).
For each arg_name value being passed, the names are evaluated for
conflicting names. In the event that a name is reused, the
decorator will fail with a `ValueError`.

Because this is part of the newly introduced arg_name behavior, there
is no backwards compatibility concern.

New tests demonstrate that applying the decorator is sufficient to
trigger the errors (you do not need to call the decorated views for us
to detect the duplicate names).

sirosen committed:
Fix grammatical error in advanced doc

Co-authored-by: Jérôme Lafréchoux <jlafrechoux@nobatek.inef4.com>

sirosen commented:
In [41e07f1](https://github.com/marshmallow-code/webargs/pull/833/commits/41e07f13eb6592253188896653e58853bbe41eb6), I introduced detection for the name conflict case which can come from stacked decorators.

It's important to note that this starts recording information in a list, `__webargs_argnames__`, which is attached to the decorated callable. It's a new side-effect for `use_args` to have, but it follows a pattern I've seen a few other libraries use for tracking information about decorators -- I'm relatively confident that it's safe given that `click` does this. Since it's part of an opt-in feature, there isn't a worry that this could make someone's existing usage start to fail unexpectedly (e.g., if they are decorating some custom callable with `__slots__`).

lafrech commented:
Couldn't this be considered a breaking change in this case?

```py
        @parser.use_args(Schema_1(partial=True))
        @parser.use_args(Schema_2(partial=True))
        def viewfunc(*, j, q):
            return (j, q)
```

I believe this used to pass but will now break.

(I don't mind introducing a breaking change and moving straight to next major.)

lafrech commented on src/webargs/core.py:
```diff
@@ -51,6 +51,12 @@
+def _record_arg_name(f: typing.Callable[..., typing.Any], argname: str) -> None:
+    if not hasattr(f, "__webargs_argnames__"):
+        f.__webargs_argnames__ = []  # type: ignore[attr-defined]
+    f.__webargs_argnames__.append(argname)  # type: ignore[attr-defined]
```
There's a (admittedly twisted) use case I try to support in flask-smorest when stacking decorators.

Assume a view is decorated once, then used in different places with other decorators (this can be achieved by calling the decorators as normal functions, not the decorator form). Then mutating the list is wrong. It should be replaced instead.

This might not make much sense in the `use_args` case and the likelihood of someone ever doing this is close to zero but it is not forbidden...

sirosen replied:
I had the idea yesterday that if we "just switched to a tuple" this use-case would fall out naturally from that.
When I went to do it today, I found that the only additional tweak which was needed was to put the `__webargs_argnames__` tuple on the `wrapper` built by the decorator, so I think this worked out quite well.

I've added a relevant test-case and it seems pretty good.

I wouldn't necessarily _advertise_ the feature because it's a bit unusual, but it is now supported and tested! :grin:

sirosen committed:
Support "multi stacking" of use_args with arg_name

Given a view which is decorated once with use_args (e.g. with query
params), it can now be decorated multiple times to produce different
variants by calling use_args on the result, without mutating shared
underlying state (a `__webargs_argnames__` list).

Now, the wrapper built by the decorator, not the decorated function,
stores `__webargs_argnames__`. Furthermore, it is stored as a tuple,
not a list, so that there is a reduced risk that it is a shared
object which is mutated unexpectedly between usages.

A test exercises this mode of usage.

sirosen commented:
> Couldn't this be considered a breaking change in this case?
> [stacked use_args usage follows]

I think we're safe because the parser option for this controls whether or not `arg_name` is even generated inside of `use_args`, and therefore whether or not it is recorded in `__webargs_argnames__`.
Without `USE_ARGS_POSITIONAL = False` on the class, `arg_name` is `None` and `__webargs_argnames__` will never be populated.

I think we're good to release this in 8.x -- which I like better anyway, since there are other changes I'd want to include in 9.0 like removing AsyncParser from the class hierarchy and I haven't thought hard about what else we should change.

lafrech commented:
> I think we're safe because the parser option for this controls whether or not `arg_name` is even generated inside of `use_args`, and therefore whether or not it is recorded in `__webargs_argnames__`.
Without `USE_ARGS_POSITIONAL = False` on the class, `arg_name` is `None` and `__webargs_argnames__` will never be populated.

That's the part I was missing from the beginning.

I guess this is good to go, then. Please allow a little time for a last look and I'll merge.

Thanks.

lafrech reviewed (approved):
Great, thanks.

I'll leave you the merge/release in case you'd like to edit the commit history before merging. But It is fine as is.

sirosen commented:
I took care of the v8.3.0 release including this this morning. From my testing it looks good.
